File: brain/src/service/persona_registry.py
"""PersonaRegistry — cached metadata about installed personas.

Answers: what personas exist, what tier, is it locked.
Does NOT answer: where content should go (that's PersonaRoutingPolicy).

Queries Core's GET /v1/personas at startup, caches the result.
Refreshes on persona-related 404, explicit events, or periodic poll.

No alias logic here — aliases are owned by PersonaRoutingPolicy.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any

import structlog

log = structlog.get_logger(__name__)

_PERSONA_PREFIX = "persona-"

# Conservative fallback when Core is unreachable at startup.
# Matches Core's bootstrap personas — not a hard invariant.
_FALLBACK_PERSONAS = [
    {"id": "persona-general", "name": "general", "tier": "default", "locked": False},
    {"id": "persona-work", "name": "work", "tier": "standard", "locked": False},
    {"id": "persona-health", "name": "health", "tier": "sensitive", "locked": True},
    {"id": "persona-finance", "name": "finance", "tier": "sensitive", "locked": True},
]


@dataclass(frozen=True)
class PersonaInfo:
    """Immutable snapshot of a persona's metadata."""

    id: str       # "persona-general"
    name: str     # "general"
    tier: str     # "default", "standard", "sensitive", "locked"
    locked: bool  # whether the vault is currently locked


class PersonaRegistry:
    """Dynamic persona registry — queries Core, caches locally.

    Thread-safe. Constructed once in main.py, injected into services.
    """

    def __init__(self) -> None:
        self._personas: dict[str, PersonaInfo] = {}
        self._loaded: bool = False
        self._lock = asyncio.Lock()

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def load(self, core: Any) -> None:
        """Initial load from Core. Falls back to defaults ONLY if no prior cache."""
        async with self._lock:
            try:
                raw = await core.list_personas_detailed()
                self._ingest(raw)
                self._loaded = True
                log.info(
                    "persona_registry.loaded",
                    count=len(self._personas),
                    names=list(self._personas.keys()),
                )
            except Exception as exc:
                if not self._personas:
                    # First load — no prior cache, use fallback
                    log.warning(
                        "persona_registry.bootstrap_fallback",
                        error=str(exc),
                        fallback_count=len(_FALLBACK_PERSONAS),
                    )
                    self._ingest(_FALLBACK_PERSONAS)
                else:
                    # Refresh failure — keep last known good cache
                    log.warning(
                        "persona_registry.refresh_failed_keeping_cache",
                        error=str(exc),
                        cached_count=len(self._personas),
                    )
                self._loaded = False

    async def refresh(self, core: Any) -> None:
        """Re-fetch from Core. Keeps last known good cache on failure."""
        await self.load(core)

    # ------------------------------------------------------------------
    # Queries (synchronous — read from cache)
    # ------------------------------------------------------------------

    def normalize(self, name: str) -> str:
        """Strip the persona- prefix added by Core."""
        if name.startswith(_PERSONA_PREFIX):
            return name[len(_PERSONA_PREFIX):]
        return name

    def exists(self, name: str) -> bool:
        """Check if a canonical persona name exists in Core."""
        return self.normalize(name) in self._personas

    def tier(self, name: str) -> str | None:
        """Return the tier for a persona (must be canonical name)."""
        info = self._personas.get(self.normalize(name))
        return info.tier if info else None

    def locked(self, name: str) -> bool | None:
        """Return whether a persona is locked (must be canonical name)."""
        info = self._personas.get(self.normalize(name))
        return info.locked if info else None

    def all_names(self) -> list[str]:
        """Return all known canonical persona names."""
        return list(self._personas.keys())

    def is_loaded(self) -> bool:
        """True if registry has been populated from Core (not fallback)."""
        return self._loaded

    def update_locked(self, name: str, locked: bool) -> None:
        """Update the locked state for a specific persona (event-driven)."""
        norm = self.normalize(name)
        info = self._personas.get(norm)
        if info:
            # PersonaInfo is frozen — replace it
            self._personas[norm] = PersonaInfo(
                id=info.id, name=info.name, tier=info.tier, locked=locked,
            )

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _ingest(self, raw_details: list[dict]) -> None:
        """Parse raw persona_details dicts into PersonaInfo objects."""
        self._personas.clear()
        for d in raw_details:
            name = d.get("name", "")
            if not name:
                raw_id = d.get("id", "")
                name = raw_id[len(_PERSONA_PREFIX):] if raw_id.startswith(_PERSONA_PREFIX) else raw_id
            if name:
                self._personas[name] = PersonaInfo(
                    id=d.get("id", f"{_PERSONA_PREFIX}{name}"),
                    name=name,
                    tier=d.get("tier", "default"),
                    locked=d.get("locked", False),
                )

File: apps/home-node-lite/core-server/src/brain/persona_registry.ts
/**
 * Task 5.44 — persona registry.
 *
 * Cached metadata about installed personas. Answers: what personas
 * exist, what tier, is it locked, what's the classification hint.
 * Does NOT answer: where content should go (that's routing policy).
 *
 * **Ownership**:
 *   - Core is the source of truth — it owns the vault files + the
 *     config that declares tiers + lock state.
 *   - Brain queries Core at startup, caches the result, refreshes
 *     periodically (60s, task 5.13) or on demand after a 404 / event.
 *   - This module is Brain's local cache — it never touches SQLite.
 *
 * **Fallback**: if Core is unreachable on first load, fall back to
 * a conservative set that matches Core's bootstrap personas. This
 * lets the classifier run against the expected-minimum set while
 * Core comes up.
 *
 * **Refresh semantics** (pinned by tests):
 *   - `refresh()` failure after a successful load keeps the last
 *     known-good cache. A transient network blip doesn't flush the
 *     cache to empty.
 *   - First-load failure installs the fallback set.
 *   - Concurrent `refresh()` calls coalesce — only one fetch fires.
 *
 * **Prefix normalisation**: Core returns persona ids like
 * `persona-general`; the registry strips the `persona-` prefix
 * internally + exposes canonical names (`general`).
 *
 * Source: docs/HOME_NODE_LITE_TASKS.md Phase 5e task 5.44.
 */

export type PersonaTier = 'default' | 'standard' | 'sensitive' | 'locked';

export interface PersonaInfo {
  /** Core's id shape (e.g. "persona-general"). */
  id: string;
  /** Canonical name (prefix stripped). */
  name: string;
  tier: PersonaTier;
  locked: boolean;
  /** Optional classification hint for PersonaSelector. */
  description: string;
}

/** Raw persona description Core returns from `GET /v1/personas`. */
export interface RawPersonaDetail {
  id?: string;
  name?: string;
  tier?: string;
  locked?: boolean;
  description?: string;
}

/** Fetcher — Brain wires this to `coreClient.list_personas_detailed()`. */
export type PersonaFetchFn = () => Promise<RawPersonaDetail[]>;

export interface PersonaRegistryOptions {
  fetchFn: PersonaFetchFn;
  /** Diagnostic hook. */
  onEvent?: (event: PersonaRegistryEvent) => void;
}

export type PersonaRegistryEvent =
  | { kind: 'loaded'; count: number; names: string[] }
  | { kind: 'fallback_used'; error: string }
  | { kind: 'refresh_failed_kept_cache'; error: string; cachedCount: number }
  | { kind: 'lock_state_changed'; name: string; locked: boolean };

const PERSONA_PREFIX = 'persona-';
const VALID_TIERS: ReadonlySet<PersonaTier> = new Set([
  'default',
  'standard',
  'sensitive',
  'locked',
]);

/**
 * Conservative fallback used only when Core is unreachable on FIRST
 * load. Matches Core's bootstrap personas.
 */
export const FALLBACK_PERSONAS: ReadonlyArray<RawPersonaDetail> = [
  { id: 'persona-general', name: 'general', tier: 'default', locked: false },
  { id: 'persona-work', name: 'work', tier: 'standard', locked: false },
  { id: 'persona-health', name: 'health', tier: 'sensitive', locked: true },
  { id: 'persona-finance', name: 'finance', tier: 'sensitive', locked: true },
];

export class PersonaRegistry {
  private readonly fetchFn: PersonaFetchFn;
  private readonly onEvent?: (event: PersonaRegistryEvent) => void;
  private readonly personas: Map<string, PersonaInfo> = new Map();
  private loaded = false;
  /** Coalesces concurrent load/refresh calls to one underlying fetch. */
  private inFlight: Promise<void> | null = null;

  constructor(opts: PersonaRegistryOptions) {
    if (typeof opts.fetchFn !== 'function') {
      throw new TypeError('PersonaRegistry: fetchFn is required');
    }
    this.fetchFn = opts.fetchFn;
    this.onEvent = opts.onEvent;
  }

  /**
   * Initial load. Falls back to `FALLBACK_PERSONAS` only if nothing
   * was cached yet. Concurrent calls coalesce.
   */
  async load(): Promise<void> {
    if (this.inFlight) return this.inFlight;
    const promise = this.doLoad();
    this.inFlight = promise;
    try {
      await promise;
    } finally {
      this.inFlight = null;
    }
  }

  /** Synonym for `load()` — kept for call-site clarity. */
  async refresh(): Promise<void> {
    return this.load();
  }

  private async doLoad(): Promise<void> {
    try {
      const raw = await this.fetchFn();
      this.ingest(Array.isArray(raw) ? raw : []);
      this.loaded = true;
      this.emit({
        kind: 'loaded',
        count: this.personas.size,
        names: Array.from(this.personas.keys()),
      });
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (this.personas.size === 0) {
        // First load — install the fallback set so the system boots.
        this.ingest(FALLBACK_PERSONAS);
        this.emit({ kind: 'fallback_used', error: msg });
      } else {
        // Refresh failure after an earlier success — keep last known cache.
        this.emit({
          kind: 'refresh_failed_kept_cache',
          error: msg,
          cachedCount: this.personas.size,
        });
      }
      this.loaded = false;
    }
  }

  // ── Query surface ────────────────────────────────────────────────────

  /** Strip the `persona-` prefix Core attaches. */
  normalize(name: string): string {
    if (typeof name !== 'string') return '';
    return name.startsWith(PERSONA_PREFIX)
      ? name.slice(PERSONA_PREFIX.length)
      : name;
  }

  exists(name: string): boolean {
    return this.personas.has(this.normalize(name));
  }

  tier(name: string): PersonaTier | null {
    return this.personas.get(this.normalize(name))?.tier ?? null;
  }

  locked(name: string): boolean | null {
    const info = this.personas.get(this.normalize(name));
    return info ? info.locked : null;
  }

  description(name: string): string {
    return this.personas.get(this.normalize(name))?.description ?? '';
  }

  allNames(): string[] {
    return Array.from(this.personas.keys());
  }

  /**
   * Return every persona's full info snapshot. Copies so callers
   * can't mutate the cache.
   */
  snapshot(): PersonaInfo[] {
    return Array.from(this.personas.values()).map((p) => ({ ...p }));
  }

  /** True only when a real Core response has loaded — false for fallback. */
  isLoaded(): boolean {
    return this.loaded;
  }

  /**
   * Event-driven lock-state update — Core can push persona-lock
   * events on the WS; no need to refetch the whole list for a single
   * flag flip.
   */
  updateLocked(name: string, locked: boolean): void {
    const norm = this.normalize(name);
    const info = this.personas.get(norm);
    if (!info) return;
    this.personas.set(norm, { ...info, locked });
    this.emit({ kind: 'lock_state_changed', name: norm, locked });
  }

  // ── Internals ────────────────────────────────────────────────────────

  private ingest(raw: ReadonlyArray<RawPersonaDetail>): void {
    this.personas.clear();
    for (const d of raw) {
      const id = typeof d.id === 'string' ? d.id : '';
      let name = typeof d.name === 'string' ? d.name : '';
      if (name === '') {
        name = id.startsWith(PERSONA_PREFIX)
          ? id.slice(PERSONA_PREFIX.length)
          : id;
      }
      if (name === '') continue; // Anonymous entry — skip
      const tier: PersonaTier = VALID_TIERS.has(d.tier as PersonaTier)
        ? (d.tier as PersonaTier)
        : 'default';
      this.personas.set(name, {
        id: id !== '' ? id : `${PERSONA_PREFIX}${name}`,
        name,
        tier,
        locked: d.locked === true,
        description: typeof d.description === 'string' ? d.description : '',
      });
    }
  }

  private emit(event: PersonaRegistryEvent): void {
    this.onEvent?.(event);
  }
}
